# Supervised Learning Model Evaluation Lab

Complete the exercises below to solidify your knowledge and understanding of supervised learning model evaluation.

In [1]:
import pandas as pd

## Regression Model Evaluation

In [190]:
from sklearn.datasets import load_boston

data = load_boston()

X = pd.DataFrame(data["data"], columns=data["feature_names"])
y = pd.DataFrame(data["target"], columns=['MEDV'])

data = pd.concat([X, y], axis=1)

In [191]:
data.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [192]:
data.shape

(506, 14)

In [193]:
data.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

## 1. Split this data set into training (80%) and testing (20%) sets.

The `MEDV` field represents the median value of owner-occupied homes (in $1000's) and is the target variable that we will want to predict.

In [6]:
from sklearn.model_selection import train_test_split

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    test_size=0.20, 
                                                    random_state=123)

## 2. Train a `LinearRegression` model on this data set and generate predictions on both the training and the testing set.

In [8]:
from sklearn.linear_model import LinearRegression

In [9]:
lr = LinearRegression()

In [171]:
lr.fit(X_train, y_train)

LinearRegression()

In [175]:
train_prediction_medv = lr.predict(X_train)
train_prediction_medv[:10]

array([[ 1.97049507],
       [ 2.07712637],
       [ 2.29083637],
       [ 0.86574276],
       [ 0.05638401],
       [-0.07214104],
       [ 1.91380713],
       [-0.04037102],
       [ 1.34004951],
       [-0.05576482]])

In [176]:
test_prediction_medv = lr.predict(X_test)
test_prediction_medv[:10]

array([[ 1.94866804],
       [ 1.83500072],
       [ 1.91952715],
       [ 1.19757304],
       [ 1.14976456],
       [ 1.23454844],
       [ 1.19166228],
       [-0.16468853],
       [ 1.30229601],
       [ 2.15665017]])

## 3. Calculate and print R-squared for both the training and the testing set.

In [14]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error, mean_squared_log_error

#### Test

In [178]:
r2_score(
    y_true = y_test,
    y_pred = test_prediction_medv
)

0.873969809031713

#### Train

In [180]:
r2_score(
    y_true = y_train,
    y_pred = train_prediction_medv
)

0.9394429054589587

## 4. Calculate and print mean squared error for both the training and the testing set.

#### Test

In [181]:
mean_squared_error(
    y_true = y_test,
    y_pred = test_prediction_medv
)

0.06287506193862322

#### Train

In [182]:
mean_squared_error(
    y_true = y_train,
    y_pred = train_prediction_medv
)

0.04289040327945002

## 5. Calculate and print mean absolute error for both the training and the testing set.

#### Test

In [184]:
mean_absolute_error(
    y_true = y_test,
    y_pred = test_prediction_medv
)

0.19220784520508186

#### Train

In [185]:
mean_absolute_error(
    y_true = y_train,
    y_pred = train_prediction_medv
)

0.16069281354685305

## Classification Model Evaluation

In [129]:
from sklearn.datasets import load_iris

data = load_iris()

X = pd.DataFrame(data["data"], columns=data["feature_names"])
y = pd.DataFrame(data["target"], columns=["class"])

data = pd.concat([X, y], axis=1)

In [130]:
data.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   class  
0      0  
1      0  
2      0  
3      0  
4      0

In [131]:
data.shape

(150, 5)

In [188]:
data.isnull().sum()

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
class                0
dtype: int64

## 6. Split this data set into training (80%) and testing (20%) sets.

The `class` field represents the type of flower and is the target variable that we will want to predict.

In [132]:
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    test_size=0.20, 
                                                    random_state=78)

## 7. Train a `LogisticRegression` model on this data set and generate predictions on both the training and the testing set.

In [133]:
from sklearn.linear_model import LogisticRegression

In [134]:
log = LogisticRegression(max_iter=10000)

In [135]:
# He puesto ".values.ravel()" porque sino me salia un warning 
log.fit(X_train, y_train.values.ravel())

LogisticRegression(max_iter=10000)

In [136]:
train_prediction_class = log.predict(X_train)
test_prediction_class = log.predict(X_test)

## 8. Calculate and print the accuracy score for both the training and the testing set.

#### Test

In [170]:
# le hemos dicho que sea el 20% de nuestros datos por eso el array tiene menos numeros que en el train 

test_prediction_class

array([2, 2, 2, 1, 1, 1, 1, 0, 1, 2, 2, 2, 2, 0, 1, 1, 1, 1, 0, 0, 0, 2,
       2, 0, 1, 1, 1, 0, 1, 2])

In [145]:
accuracy_score(
    y_true = y_test,
    y_pred = test_prediction_class
)

0.9333333333333333

#### Train

In [144]:
train_prediction_class

array([2, 2, 2, 1, 0, 0, 2, 0, 1, 0, 1, 2, 2, 1, 2, 0, 0, 0, 2, 0, 2, 0,
       1, 0, 1, 1, 1, 0, 0, 0, 1, 2, 1, 0, 1, 0, 2, 2, 1, 1, 2, 2, 0, 0,
       2, 1, 0, 0, 0, 1, 1, 1, 2, 0, 2, 2, 2, 2, 2, 2, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 2, 0, 2, 2, 2, 0, 0, 1, 2, 1, 2, 2, 1, 1, 1, 0, 0,
       2, 2, 1, 2, 0, 0, 1, 1, 2, 0, 1, 0, 2, 2, 1, 1, 0, 2, 0, 2, 1, 0,
       2, 0, 0, 1, 2, 2, 0, 1, 2, 0])

In [146]:
accuracy_score(
    y_true = y_train,
    y_pred = train_prediction_class
)

0.9916666666666667

## 9. Calculate and print the balanced accuracy score for both the training and the testing set.

In [69]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, fbeta_score, balanced_accuracy_score

In [72]:
# sklearn.metrics.balanced_accuracy_score(y_true, y_pred, *, sample_weight=None, adjusted=False)
# basandome en esta funcion de la documentacion de sklearn

#### Test

In [148]:
balanced_accuracy_score(y_true = y_test,
                        y_pred = test_prediction_class,
                        sample_weight=None,
                        adjusted=False)

0.9555555555555556

#### Train

In [149]:
balanced_accuracy_score(y_true = y_train,
                        y_pred = train_prediction_class,
                        sample_weight=None,
                        adjusted=False)

0.9904761904761905

## 10. Calculate and print the precision score for both the training and the testing set.

#### Test

In [151]:
precision_score(
    y_true = y_test,
    y_pred = test_prediction_class,
    average = None)


array([1. , 1. , 0.8])

#### Train

In [152]:
precision_score(
    y_true = y_train,
    y_pred = train_prediction_class,
    average = None)


array([1.        , 1.        , 0.97674419])

## 11. Calculate and print the recall score for both the training and the testing set.

#### Test

In [155]:
recall_score(
    y_true = y_test,
    y_pred = test_prediction_class,
    average = None
)

array([1.        , 0.86666667, 1.        ])

#### Train

In [157]:
recall_score(
    y_true = y_train,
    y_pred = train_prediction_class,
    average = None)

array([1.        , 0.97142857, 1.        ])

## 12. Calculate and print the F1 score for both the training and the testing set.

#### Test

In [163]:
f1_score(
    y_true = y_test,
    y_pred = test_prediction_class,
    average = None
)

array([1.        , 0.92857143, 0.88888889])

#### Train

In [165]:
f1_score(
    y_true = y_train,
    y_pred = train_prediction_class,
    average = None
)

array([1.        , 0.98550725, 0.98823529])

## 13. Generate confusion matrices for both the training and the testing set.

In [89]:
from sklearn.metrics import confusion_matrix

#### Test

In [168]:
confusion_matrix(
    y_true = y_test,
    y_pred = test_prediction_class
)

array([[ 7,  0,  0],
       [ 0, 13,  2],
       [ 0,  0,  8]], dtype=int64)

#### Train

In [169]:
confusion_matrix(
    y_true = y_train,
    y_pred = train_prediction_class
)

array([[43,  0,  0],
       [ 0, 34,  1],
       [ 0,  0, 42]], dtype=int64)

## Bonus: For each of the data sets in this lab, try training with some of the other models you have learned about, recalculate the evaluation metrics, and compare to determine which models perform best on each data set.

### 1. BOSTON

In [195]:
from sklearn.datasets import load_boston

data = load_boston()

X = pd.DataFrame(data["data"], columns=data["feature_names"])
y = pd.DataFrame(data["target"], columns=['MEDV'])

data = pd.concat([X, y], axis=1)

In [196]:
data.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

#### Voy a practicar el modelo Gridsearch

In [245]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Ridge
import numpy as np

In [246]:
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    test_size=0.20, 
                                                    random_state=123)

In [249]:
model = Ridge()
parameter_space = {'alpha': np.linspace(0, 100, 1000),
                   }

grid_search = GridSearchCV(model,
                           param_grid=parameter_space,
                           cv=3)

grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, estimator=Ridge(),
             param_grid={'alpha': array([  0.        ,   0.1001001 ,   0.2002002 ,   0.3003003 ,
         0.4004004 ,   0.5005005 ,   0.6006006 ,   0.7007007 ,
         0.8008008 ,   0.9009009 ,   1.001001  ,   1.1011011 ,
         1.2012012 ,   1.3013013 ,   1.4014014 ,   1.5015015 ,
         1.6016016 ,   1.7017017 ,   1.8018018 ,   1.9019019 ,
         2.002002  ,   2.1021021 ,   2.2022022 ,   2.3023023 ,
         2.4024024 ,   2.5025025 ,   2.6026026 ,   2.7027027 ,
         2.8028028 ,   2.9029...
        96.4964965 ,  96.5965966 ,  96.6966967 ,  96.7967968 ,
        96.8968969 ,  96.996997  ,  97.0970971 ,  97.1971972 ,
        97.2972973 ,  97.3973974 ,  97.4974975 ,  97.5975976 ,
        97.6976977 ,  97.7977978 ,  97.8978979 ,  97.997998  ,
        98.0980981 ,  98.1981982 ,  98.2982983 ,  98.3983984 ,
        98.4984985 ,  98.5985986 ,  98.6986987 ,  98.7987988 ,
        98.8988989 ,  98.998999  ,  99.0990991 ,  99.1991992 ,
        99.29

In [250]:
best_model = grid_search.best_estimator_
best_model

Ridge(alpha=1.001001001001001)

In [251]:
y_pred = best_model.predict(X_test)

In [252]:
r2_score(y_test, y_pred)

0.9369312459614665

In [253]:
grid_search.best_score_

0.9150771822971983

In [254]:
mean_squared_error( y_test, y_pred)

0.05017469765732219

In [255]:
mean_absolute_error( y_test, y_pred)

0.1716069607644792

#### Conclusiones: 
1. El best score nos indica el porcentaje de precision de nuestro modelo que es un 91.5 % (no está nada mal)
2. El r2 ha mejorado en este método que en el método de regresión lineal utilizado antes --> ha aumentado
3. El mean aquared error tambien ha mejorado mediante el model de gridge, ha disminuido una centesima en comparacion con el de regresion lineal
4. Y, el mean absoulute erros también ha mejorado porque ha disminutido mediante este modelo 
5. Podemos decir que el Modelo de Gridge es más preciso que el modelo de Regresión lineal

### 2. IRIS

In [215]:
from sklearn.datasets import load_iris

data = load_iris()

X = pd.DataFrame(data["data"], columns=data["feature_names"])
y = pd.DataFrame(data["target"], columns=["class"])

data = pd.concat([X, y], axis=1)

In [216]:
data.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   class  
0      0  
1      0  
2      0  
3      0  
4      0

#### Voy a pracicar el modelo de KNeighborsClassifier

In [217]:
from sklearn.neighbors import KNeighborsClassifier

In [219]:
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    test_size=0.20, 
                                                    random_state=78)

In [229]:
n_neighbors = np.arange(1,50)


model = KNeighborsClassifier()
parameter_space = {'n_neighbors': n_neighbors,
                   }

grid_search = GridSearchCV(model,
                           param_grid=parameter_space,
                           cv=3)

grid_search.fit(X_train, y_train.values.ravel())

GridSearchCV(cv=3, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])})

In [230]:
grid_search.best_score_

0.9833333333333334

In [231]:
grid_search.best_estimator_

KNeighborsClassifier(n_neighbors=7)

#### Conclusiones:
1. El best score nos indica el procentaje de precision de nuestro modelo y este es un 98.3% (la verdad es que está fenomenal)
2. En comparación con las métricas utilizadas en en modelo de Logistica, este presenta una precisión mejor asique me quedaría con este modelo de clasificación

### DUDAS
1. No llego a entender como determinamos el "random_state" cuando hacemos el train-test splitting (nos lo inventamos... no nos lo inventamos...)
2. Tampoco entiendo por qué la "matriz de confusión" me da una matriz de 3x3, creía haber entendido que nos devolvia una matriz 2x2 que nos decia que predicciones habiamos acertado y cuales no
3. No consigo entender como determinamos el "parameter_space" del modelo Gridsearch (nos lo inventamos... no nos lo inventamos...)
4. Y tampoco consigo entender como determinamos el "parameter_space" del modelo KNeighborsClassifier (nos lo inventamos... no nos lo inventamos...)

* Creo que tengo un problema con los espacios... que me fallan!! jajaja